File: app/src/main/java/org/baiyu/fuckshare/filetype/FileType.java
package org.baiyu.fuckshare.filetype;

import java.nio.ByteBuffer;
import java.util.Comparator;
import java.util.Map;
import java.util.Set;

public interface FileType {
    String getExtension();

    Set<Map<Integer, byte[]>> getSignatures();

    default boolean isSupportMetadata() {
        return false;
    }

    default boolean signatureMatch(byte[] bytes) {
        Set<Map<Integer, byte[]>> signature = getSignatures();
        if (bytes == null || signature == null) {
            return false;
        }

        int signatureRequireLength = signature.parallelStream()
                .flatMap(m -> m.entrySet().stream())
                .map(integerEntry -> integerEntry.getKey() + integerEntry.getValue().length)
                .max(Comparator.comparingInt(x -> x))
                .orElse(Integer.MAX_VALUE);

        if (signatureRequireLength <= 0 || bytes.length < signatureRequireLength) {
            return false;
        }

        return signature.parallelStream().anyMatch(
                integerMap -> integerMap.entrySet().parallelStream().allMatch(
                        integerEntry -> ByteBuffer.wrap(integerEntry.getValue())
                                .equals(ByteBuffer.wrap(bytes, integerEntry.getKey(), integerEntry.getValue().length))
                )
        );
    }
}

File: app/src/main/java/org/baiyu/fuckshare/filetype/FileType.kt
package org.baiyu.fuckshare.filetype

import java.nio.ByteBuffer

interface FileType {
    val extension: String?
    val signatures: Set<Map<Int, ByteArray>>?
    val supportMetadata: Boolean
        get() = false

    fun signatureMatch(bytes: ByteArray): Boolean {
        return signatures?.parallelStream()?.anyMatch { integerMap ->
            integerMap.entries.parallelStream()
                .allMatch {
                    bytes.size >= it.key + it.value.size
                            && ByteBuffer.wrap(it.value)
                        .equals(ByteBuffer.wrap(bytes, it.key, it.value.size))
                }
        } ?: false
    }
}